File: Q/Encryption/StringEncryption.vb
Imports System.Security.Cryptography
Imports System.IO
Imports System.Text

Public Class StringEncryption

    Private _encryptionKey As String
    Private ReadOnly bytIV() As Byte = {121, 9, 10, 1, 31, 74, 11, 39, 255, 91, 45, 78, 55, 66, 22, 7}

    Public Sub New(ByVal encryptionKey As String)
        _encryptionKey = encryptionKey
    End Sub

    ''' <summary>
    ''' Encrypts text
    ''' </summary>
    ''' <param name="textToBeEncrypted">Text to be encrypted</param>
    ''' <returns>Encrypted text</returns>
    ''' <remarks></remarks>
    Public Function EncryptString(ByVal textToBeEncrypted As String) As String
        Dim bytValue() As Byte
        Dim bytKey() As Byte
        Dim bytEncoded() As Byte = New Byte() {}
        Dim intLength As Integer
        Dim intRemaining As Integer
        Using objMemoryStream As New MemoryStream
            Dim objCryptoStream As CryptoStream
            Using objRijndaelManaged As New RijndaelManaged
                'Strip any null character from string to be encrypted
                textToBeEncrypted = StripNullCharacters(textToBeEncrypted)
                'Value must be within ASCII range
                bytValue = Encoding.ASCII.GetBytes(textToBeEncrypted.ToCharArray)
                intLength = Len(_encryptionKey)
                '   ******   Encryption Key must be 256 bits long (32 bytes)      ******
                '   ******   If it is longer than 32 bytes it will be truncated.  ******
                '   ******   If it is shorter than 32 bytes it will be padded     ******
                '   ******   with upper-case Xs.                                  ****** 
                If intLength >= 32 Then
                    _encryptionKey = Strings.Left(_encryptionKey, 32)
                Else
                    intLength = Len(_encryptionKey)
                    intRemaining = 32 - intLength
                    _encryptionKey = _encryptionKey & Strings.StrDup(intRemaining, "X")
                End If
                bytKey = Encoding.ASCII.GetBytes(_encryptionKey.ToCharArray)
                'Create the encryptor and write value to it after it is converted into a byte array
                Try
                    objCryptoStream = New CryptoStream(objMemoryStream, objRijndaelManaged.CreateEncryptor(bytKey, bytIV), CryptoStreamMode.Write)
                    objCryptoStream.Write(bytValue, 0, bytValue.Length)
                    objCryptoStream.FlushFinalBlock()
                    bytEncoded = objMemoryStream.ToArray()
                    objMemoryStream.Close()
                    objCryptoStream.Close()
                Catch
                End Try
            End Using
        End Using
        'Return encrypted value (converted from  byte Array to a base64 string).  Base64 is MIME encoding)
        Return Convert.ToBase64String(bytEncoded)
    End Function

    ''' <summary>
    ''' Decrypts text
    ''' </summary>
    ''' <param name="textToBeDecrypted">Text to be decrypted</param>
    ''' <returns>Decrypted text</returns>
    ''' <remarks></remarks>
    Public Function DecryptString(ByVal textToBeDecrypted As String) As String
        Dim bytDataToBeDecrypted() As Byte
        Dim bytTemp() As Byte
        Using objRijndaelManaged As New RijndaelManaged
            Dim objMemoryStream As MemoryStream
            Dim objCryptoStream As CryptoStream
            Dim bytDecryptionKey() As Byte

            Dim intLength As Integer
            Dim intRemaining As Integer
            Dim strReturnString As String = String.Empty
            '   ******   Convert base64 encrypted value to byte array      ******
            bytDataToBeDecrypted = Convert.FromBase64String(textToBeDecrypted)
            '   ******   Encryption Key must be 256 bits long (32 bytes)      ******
            '   ******   If it is longer than 32 bytes it will be truncated.  ******
            '   ******   If it is shorter than 32 bytes it will be padded     ******
            '   ******   with upper-case Xs.                                  ****** 
            intLength = Len(_encryptionKey)
            If intLength >= 32 Then
                _encryptionKey = Strings.Left(_encryptionKey, 32)
            Else
                intLength = Len(_encryptionKey)
                intRemaining = 32 - intLength
                _encryptionKey = _encryptionKey & Strings.StrDup(intRemaining, "X")
            End If
            bytDecryptionKey = Encoding.ASCII.GetBytes(_encryptionKey.ToCharArray)
            ReDim bytTemp(bytDataToBeDecrypted.Length)
            objMemoryStream = New MemoryStream(bytDataToBeDecrypted)
            '   ******  Create the decryptor and write value to it after it is   ******
            '   ******  converted into a byte array                          ******
            Try
                objCryptoStream = New CryptoStream(objMemoryStream,
               objRijndaelManaged.CreateDecryptor(bytDecryptionKey, bytIV), CryptoStreamMode.Read)
                objCryptoStream.Read(bytTemp, 0, bytTemp.Length)
                objCryptoStream.FlushFinalBlock()
                objMemoryStream.Close()
                objCryptoStream.Close()
            Catch
            End Try
        End Using
        '   ******   Return decypted value     ******
        Return StripNullCharacters(Encoding.ASCII.GetString(bytTemp))

    End Function

    Private Function StripNullCharacters(ByVal vstrStringWithNulls As String) As String
        Dim intPosition As Integer
        Dim strStringWithOutNulls As String
        intPosition = 1
        strStringWithOutNulls = vstrStringWithNulls
        Do While intPosition > 0
            intPosition = InStr(intPosition, vstrStringWithNulls, vbNullChar)
            If intPosition > 0 Then
                strStringWithOutNulls = Left$(strStringWithOutNulls, intPosition - 1) &
                                  Right$(strStringWithOutNulls, Len(strStringWithOutNulls) - intPosition)
            End If
            If intPosition > strStringWithOutNulls.Length Then
                Exit Do
            End If
        Loop
        Return strStringWithOutNulls
    End Function

End Class

File: R/RegentsApp/StringEncryption.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace RegentsApp
{
    public class StringEncryption
    {
        #region String Encryption/Decryption
        private string _encryptionKey = "Regents e muito legal, mas pode dar-lhe uma ulcera ou conduzi-lo a bebida.";
        private byte[] _salt = { 121, 9, 10, 1, 31, 74, 11, 39, 255, 91, 45, 78, 55, 66, 22, 7 };

        /// <summary>
        /// Encrypts a string such that it can be decrypted using the "string.Decrypt()" method.
        /// The encryption key, salt, and encryption mechanism match those used
        /// in the internal Regents' program's "reset password" function.
        /// </summary>
        public string Encrypt(string unencryptedString)
        {
            //Remove any null characters from the unencrypted string.
            unencryptedString = unencryptedString.Replace("\0", "");
            //Ensure the encryption key is 32 bytes. Pad with 'X' characters if needed.
            if (_encryptionKey.Length > 32)
            {
                _encryptionKey = _encryptionKey.Substring(0, 32);
            }
            else
            {
                _encryptionKey = _encryptionKey.PadRight(32, 'X');
            }

            //Create the encryptor and send it a byte array of the unencrypted string.
            try
            {
                string encryptedString = string.Empty;
                using (MemoryStream memoryStream = new MemoryStream())
                {
                    byte[] keyBytes = Encoding.ASCII.GetBytes(_encryptionKey.ToCharArray());
                    byte[] unencryptedBytes = Encoding.ASCII.GetBytes(unencryptedString.ToCharArray());
                    using (CryptoStream encryptionStream = new CryptoStream(memoryStream, new RijndaelManaged().CreateEncryptor(keyBytes, _salt), CryptoStreamMode.Write))
                    {
                        encryptionStream.Write(unencryptedBytes, 0, unencryptedBytes.Length);
                        encryptionStream.FlushFinalBlock();
                        encryptionStream.Close();
                    }
                    encryptedString = Convert.ToBase64String(memoryStream.ToArray());
                    memoryStream.Close();
                }
                return encryptedString;
            }
            catch (Exception)
            {
                return string.Empty;
            }
        }//Encrypt()

        /// <summary>
        /// Decrypts a string that was encrypted using the "string.Encrypt()" method.
        /// The encryption key, salt, and encryption mechanism match those used
        /// in the internal Regents' program's "reset password" function.
        /// </summary>
        public string Decrypt(string encryptedString)
        {
            //Ensure the encryption key is 32 bytes. Pad with 'X' characters if needed.
            if (_encryptionKey.Length > 32)
            {
                _encryptionKey = _encryptionKey.Substring(0, 32);
            }
            else
            {
                _encryptionKey = _encryptionKey.PadRight(32, 'X');
            }

            //Create the decryptor and send it a byte array of the encrypted string.
            try
            {
                string decryptedString = string.Empty;
                byte[] encryptedBytes = Convert.FromBase64String(encryptedString);
                using (MemoryStream memoryStream = new MemoryStream(encryptedBytes))
                {
                    byte[] keyBytes = Encoding.ASCII.GetBytes(_encryptionKey.ToCharArray());
                    byte[] decryptedBytes = new byte[encryptedBytes.Length];
                    RijndaelManaged encrypter = new RijndaelManaged();
                    using (CryptoStream decryptionStream = new CryptoStream(memoryStream, encrypter.CreateDecryptor(keyBytes, _salt), CryptoStreamMode.Read))
                    {
                        decryptionStream.Read(decryptedBytes, 0, decryptedBytes.Length);
                        decryptionStream.Close();
                    }
                    decryptedString = Encoding.ASCII.GetString(decryptedBytes).Replace("\0", "");
                    memoryStream.Close();
                }
                return decryptedString;
            }
            catch (Exception ex)
            {
                return ex.ToString();
                //return string.Empty;
            }
        }//Decrypt()
        #endregion String Encryption/Decryption
    }//class
}//namespace